algorithms/linfa-elasticnet/src/algorithm.rs

```diff
@@ -217,22 +443,68 @@ fn duality_gap<'a, F: Float>(
     gap
 }
 
-fn variance_params<F: Float + Lapack, T: AsTargets<Elem = F>, D: Data<Elem = F>>(
-    ds: &DatasetBase<ArrayBase<D, Ix2>, T>,
-    y_est: Array1<F>,
+fn duality_gap_mtl<'a, F: Float>(
+    x: ArrayView2<'a, F>,
+    y: ArrayView2<'a, F>,
+    w: ArrayView2<'a, F>,
+    r: ArrayView2<'a, F>,
+    l1_ratio: F,
+    penalty: F,
+) -> F {
+    let half = F::cast(0.5);
+    let n_samples = F::cast(x.shape()[0]);
+    let l1_reg = l1_ratio * penalty * n_samples;
+    let l2_reg = (F::one() - l1_ratio) * penalty * n_samples;
+    let xta = x.t().dot(&r) - &w * l2_reg;
+
+    let dual_norm_xta = xta
+        .map_axis(Axis(1), |x| x.dot(&x).sqrt())
+        .fold(F::zero(), |max_norm, &nrm| max_norm.max(nrm));
+    let r_norm2 = r.fold(F::zero(), |sum, &rij| sum + rij.powi(2));
+    let w_norm2 = w.fold(F::zero(), |sum, &wij| sum + wij.powi(2));
+    let (const_, mut gap) = if dual_norm_xta > l1_reg {
+        let const_ = l1_reg / dual_norm_xta;
+        let a_norm2 = r_norm2 * const_ * const_;
+        (const_, half * (r_norm2 + a_norm2))
+    } else {
+        (F::one(), r_norm2)
+    };
+    let rty = r.t().dot(&y);
+    let trace_rty = rty.diag().sum();
+    let l21_norm = w.map_axis(Axis(1), |wj| (wj.dot(&wj)).sqrt()).sum();
+    gap += l1_reg * l21_norm - const_ * trace_rty
+        + half * l2_reg * (F::one() + const_ * const_) * w_norm2;
+    gap
+}
+
+fn variance_params<F: Float + Lapack, T: AsTargets<Elem = F>, D: Data<Elem = F>, I: Dimension>(
```

Comment:
This `AsTargets<Elem = F>` is very complex to deal with. I can't do much with it, since I need to have an `ArrayBase` in order to call `ndim()`, `shape()` and to make the computation `target - y_est`.

Reply:
Actually it should have some way to retrieve the dimension of the targets.

Reply:
Currently `AsTargets` has the method `as_multi_targets`, which returns a 2D array view, so you can call it to retrieve the target for both cases. For the single target case this returns an array of dimension (n, 1). This means your code needs to treat single-task and multi-task-with-only-one-task as equivalent cases. `y_est` will need to be a 2D array in all cases; for single-task just insert `Axis(1)` into `y_est`.

After your other PR, this should be bounded with `AsMultiTargets` (now that I think about it, we need `AsMultiTargets` as a super-trait of `AsSingleTarget` for this to work).